### PR TITLE
Basic peer scoring

### PR DESCRIPTION
**Pull request by marc-casperlabs**
Implements a basic system that will prefer to ask peers for data that previously provided some by keeping a list of offenders called "bad peers". If a peer times out or reports as not having data available, it is marked bad and not asked for data anymore.

Every 10k download attempts (this number is configurable) the old bad peer will be redeemed, i.e. removed from the bad list, giving it a chance to be included in the next round of downloads. Should the node run out of peers entirely, all bad peers will be redeemed immediately.

A refactoring of passing the context around in more functions was necessary. This hit a little hiccup at the beginning of the process, since the trusted hash is part of the context, but also fetched using functions that require it. It is left to the user of the struct to ensure that the trusted hash will not be accessed prematurely to avoid a panic.

### CODE REVIEW

File: node/src/components/chain_synchronizer/operations.rs [hunk 1]
```diff
@@ -68,54 +69,147 @@ impl<'a> Drop for ScopeTimer<'a> {
...
 where
     T: Item + 'static,
     JoinerEvent: From<FetcherRequest<T>>,
 {
     loop {
-        for peer in effect_builder.get_fully_connected_peers().await {
+        let mut new_peer_list = ctx.effect_builder.get_fully_connected_peers().await;
+        ctx.filter_bad_peers(&mut new_peer_list);
+
+        for peer in new_peer_list {
```
Comment (goral09) [inline node/src/components/chain_synchronizer/operations.rs:209]: `new_peer_list` may contain bad nodes. We should get the peers from `ctx`.

Comment (marc-casperlabs) [reply node/src/components/chain_synchronizer/operations.rs:209]: It's supposed to be this way around though. `new_peer_list` is all nodes, while we only track the bad nodes in `ctx`. Then we use the list of bad nodes in `ctx` to remove those from `new_peer_list` and are left with only the non-bad aka good ones.

File: node/src/components/chain_synchronizer/operations.rs [hunk 2]
```diff
@@ -68,54 +69,147 @@ impl<'a> Drop for ScopeTimer<'a> {
...
+    /// Marks a peer as bad.
+    fn mark_bad_peer(&self, peer: NodeId) {
+        let mut bad_peer_list = self
+            .bad_peer_list
+            .write()
+            .expect("bad peer list lock poisoned");
+
+        // Note: Like `filter_bad_peers`, this may need to be migrated to use sets instead.
+        if bad_peer_list.contains(&peer) {
+            bad_peer_list.push_back(peer);
```
Comment (marc-casperlabs) [inline node/src/components/chain_synchronizer/operations.rs]: This is the wrong way round, I'll fix it.

File: node/src/components/chain_synchronizer/operations.rs (hunk 1 above)
Comment (goral09) [reply node/src/components/chain_synchronizer/operations.rs:209]: I see. I hate mutating a value inside a function. Can we consume `new_peer_list` and return filtered out collection?

Comment (marc-casperlabs) [reply node/src/components/chain_synchronizer/operations.rs:209]: That'd needless copy the entire thing and reallocate though, whereas this allows `retain` to save the allocation (I hope). Do you think that's worth it?

(The interface mirror's `Vec::retain`)

File: node/src/components/chain_synchronizer/operations.rs [hunk 3]
```diff
@@ -68,54 +69,148 @@ impl<'a> Drop for ScopeTimer<'a> {
...
+    /// Clears the list of bad peers.
+    fn redeem_all(&self) {
+        let mut bad_peer_list = self
+            .bad_peer_list
+            .write()
+            .expect("bad peer list lock poisoned");
+
+        let bad_peer_list = mem::take::<VecDeque<NodeId>>(&mut bad_peer_list);
+        if !bad_peer_list.is_empty() {
+            warn!(?bad_peer_list, "redeemed all bad peers")
```
Comment (goral09) [inline node/src/components/chain_synchronizer/operations.rs]: Let's just log the number of bad peers instead of their ids.

Comment (marc-casperlabs) [reply node/src/components/chain_synchronizer/operations.rs]: Right, I assume we can get their IDs from the earlier logs.

**Review by goral09 — APPROVED** [review]
If I understand the code correctly, we might panic if we try to start subsequent fast-sync steps before we finish to synchronize the initial trusted block. I think we should be allowed to make this assumptions. Obviously, we would prefer to not panic at all but for the sake of tests we can continue with this.

**Pull request merged**